arch/rv32i/src/lib.rs

```diff
@@ -425,6 +425,39 @@ pub extern "C" fn _start_trap() {
     }
 }
 
+/// RISC-V semihosting needs three exact instructions in uncompressed form.
+///
+/// See https://github.com/riscv/riscv-semihosting-spec/blob/main/riscv-semihosting-spec.adoc#11-semihosting-trap-instruction-sequence
+/// for more details on the three insturctions.
+///
+/// In order to work with semihosting we include the assembly here
+/// where we are able to disable compressed instruction support. This
+/// follows the example used in the Linux kernel:
+/// https://elixir.bootlin.com/linux/v5.12.10/source/arch/riscv/include/asm/jump_label.h#L21
+/// as suggested by the RISC-V developers:
+/// https://groups.google.com/a/groups.riscv.org/g/isa-dev/c/XKkYacERM04/m/CdpOcqtRAgAJ
+#[cfg(all(target_arch = "riscv32", target_os = "none"))]
+pub unsafe fn semihost_command(_command: usize, _arg0: usize, _arg1: usize) -> ! {
```

Comment:
I don't quite understand this part. The parameters don't seem to be used so I guess this will always issue the particular semihosting command to exit the emulation?

In that case, I'd vouch to remove the arguments and make the function specific for exiting emulation.

Otherwise, if this is supposed to be a generic semihosting interface, I don't think we can / should mark this function as non-returning, because presumably some semihosting commands will continue execution at some point? In short, I see a mismatch between having the generic arguments, and the return value being specific for one particular semihosting command.

Reply:
I see that there are indeed some arguments passed into this function. If these are used because they're loaded into registers, we definitely want to use a defined (C) ABI here. If this works because the arguments are currently loaded in the correct registers, but it's not marked `extern "C"`, that is probably pure coincidence.

Reply:
Good catch, the `extern "C"` was lost in the conversion to inline asm and it continued to work. I have added it back.

I also changed `semihost_command` to not be `-> !` as you are right, some commands could return. I then added non returning helper functions for the exit commands.

Reply:
Great, this looks much better :+1:. I still think we need to remove the `noreturn` `asm!` option, otherwise this looks good to me.